### PR TITLE
#5567: Replacing System.Drawing for DesktopGL

### PR DESCRIPTION
**Pull request by rds1983**

### CODE REVIEW

**Comment by cra0zy** [comment]
MacOS/MONOMAC platform is not a DesktopGL platform, you partially modified it in here and that is why the build is failing.

**Comment by rds1983** [comment]
Thanks for the help so far.
Is there a way to see the actual error text?
Not just "Process exited with code 1".

**Comment by rds1983** [comment]
Nevermind, found error in the build log.

**Comment by dellis1972** [comment]
@rds1983  I noticed the RGBToBGR function is gone. Did you make sure that it was not needed anymore? You need to test loading from .xnb and .png files and ensure transparent images are rendering correctly. 
The reason we had this is because OpenGL expects the image in BGR format (if I remember correctly).

**Comment by rds1983** [comment]
ImageEx.cs was relying on System.Drawing. That why it had to go.
It was used only in the DesktopGL's version of Texture2D.FromStream. .xnb loading wasnt using it, so it shouldn't be affected.
I've attached a small sample that loads 2 textures using Texture2D.FromStream and then renders 2nd texture, which is transparent, on top of the first. It seems to work the same way it had before this PR(when it used Bitmap.FromStream for loading).

also the sample has commented section which could be used to test out SaveAsPng and SaveAsJpg.

[MonoGameStb.Test.zip](https://github.com/MonoGame/MonoGame/files/881775/MonoGameStb.Test.zip)

**Comment by tomspilman** [comment]
First... thanks for all your work here @rds1983 ...  big issues like this only get solved with people like you contributing your time and efforts to MonoGame.

@KonajuGames @dellis1972 @rds1983 - So if this new image code is fully managed and works on all platforms...  is it a waste of effort to just add it to DesktopGL only?  When should we purge all the `PlatformFromStream` methods and move this code into the common `FromStream` method?

Second when should we purge the previous PNG loading/saving code?

  https://github.com/MonoGame/MonoGame/tree/develop/MonoGame.Framework/Utilities/Png

**Comment by cra0zy** [comment]
> So if this new image code is fully managed and works on all platforms... is it a waste of effort to just add it to DesktopGL only? When should we purge all the PlatformFromStream methods and move this code into the common FromStream method?

IMO it's better to modify just DesktopGL in this PR so we can test it out a bit before we use it for the rest  of the platforms. It also makes it far easier to merge since it's modifying just one platform.

**Comment by rds1983** [comment]
Yes, the new code works on all platforms and even on .NET Core. I think it is good idea to use it on more platforms as it'll provide consisency of Texture2D.FromStream result across different platforms.
Right now, the result is not consisent. I.e. WindowsDX premultiplies alpha, while DesktopGL doesnt.
I think there is even [an issue](https://github.com/MonoGame/MonoGame/issues/5529) about that.
On the other hand new image code is about 300kb. So it'll definitely make MonoGame size bigger. That could be an issue for the mobile platforms.

I think removing old Png code is good idea too. Instead following classes could be made public:
[ImageReaderFromStream](https://github.com/MonoGame/MonoGame/pull/5630/files#diff-ca91ee5e1600a7737bc6ffca45d62e83)
[ImageWriterToStream](https://github.com/MonoGame/MonoGame/pull/5630/files#diff-4e2ffaf2d7fc51115e741f1cc731bc27)
Which will allow users to read images in GIF(even animated)/PNG/JPG/BMP/TGA formats. And write to PNG/JPG/TGA and BMP formats.
Though class names could be reviewed. As well as their public interfaces.

**Comment by dellis1972** [comment]
Size will only be an issue on Mobile if people use that code path.
Otherwise the linker will strip out the unused code.

**Comment by tomspilman** [comment]
> IMO it's better to modify just DesktopGL in this PR so we can test it out a bit before we use it for the rest of the platforms. 

Yeah that makes sense to me.

My only concern is we would be merging this code blind... we have unit tests for `Texture2D.FromStream` as well as `SaveToPng`.  We just don't have the unit tests for DesktopGL hooked up to the build server yet.

So someone would have to manually run the DesktopGL tests to see if this at least passes what tests we have now.

**Comment by rds1983** [comment]
I didnt change SaveToPng. It still uses Utilities/Png.
FromStream and SaveToJpg had been changed.

**Comment by tomspilman** [comment]
> I didnt change SaveToPng. It still uses Utilities/Png.

I would switch that over too so we can test it all under DesktopGL?

@KonajuGames @dellis1972 -What do you guys think?

**Comment by cra0zy** [comment]
> I would switch that over too so we can test it all under DesktopGL?

I agree.

**Comment by rds1983** [comment]
> I would switch that over too so we can test it all under DesktopGL?
I have fixed it. If anyone has objections it would be very it roll that change back.

Also, I've reviewed entire changed code again and found another place that had been affected: MouseCursor.SDL.cs, method PlatformFromTexture2D.
It uses the new code to save the image to bmp. I hope there's a corresponding test.

**Comment by rds1983** [comment]
>> I would switch that over too so we can test it all under DesktopGL?
> I have fixed it. If anyone has objections it would be very it roll that change back.

It wasn't that simple. As all platforms but android use Utilities/Png for SaveAsPng.
Means the new code imaging code needs to be added for all platforms (right now, it is added for WindowsGL,Linux,MacOS platforms). While it'll be used only for SaveAsJpg/SaveAsPng. 
Should I proceed?

**Comment by cra0zy** [comment]
> It wasn't that simple. As all platforms but android use Utilities/Png for SaveAsPng.
> Means the new code imaging code needs to be added for all platforms (right now, it is added for WindowsGL,Linux,MacOS platforms). While it'll be used only for SaveAsJpg/SaveAsPng.
> Should I proceed?

I don't get what problem are you having?

File: Build/Projects/MonoGame.Framework.definition [hunk 1]
```diff
@@ -1206,6 +1203,40 @@
     <Compile Include="Utilities\Png\PngReader.cs" />
     <Compile Include="Utilities\Png\PngWriter.cs" />
     <Compile Include="Utilities\ByteBufferPool.cs" />
+    <Compile Include="Utilities\Imaging\Image.cs">
+        <Platforms>WindowsGL,Linux,MacOS</Platforms>
```
Comment (cra0zy) [inline Build/Projects/MonoGame.Framework.definition]: MacOS shouldn't be here.

**Comment by rds1983** [comment]
Nevermind. I've realized that the solution is to make the new code be called in SaveAsPng for WindowsGL,Linux,MacOS platforms and Utilities/Png for other platforms.

File: MonoGame.Framework/Graphics/Texture2D.OpenGL.cs [hunk 2]
```diff
@@ -495,13 +479,12 @@ private void PlatformSaveAsPng(Stream stream, int width, int height)
...
 #else
-            var pngWriter = new PngWriter();
-            pngWriter.Write(this, stream);
+            SaveAsImage(stream, width, height, ImageWriterType.Png);
 #endif
         }
 
 #if DESKTOPGL || MONOMAC
-        internal void SaveAsImage(Stream stream, int width, int height, ImageFormat format)
+        internal void SaveAsImage(Stream stream, int width, int height, ImageWriterType format)
```
Comment (cra0zy) [inline MonoGame.Framework/Graphics/Texture2D.OpenGL.cs]: You modified a piece of code that is used by both DESKTOPGL and MONOMAC, correct thing would be to add `#elif DESKTOPGL` and add code bellow.

**Comment by rds1983** [comment]
> I don't get what problem are you having?
Nevermind, I've made it so new code is used solely for the DesktopGL(Texture2D.FromStream, Texture2D.SaveAsJpg and Texture2D.SaveAsPng). While other platforms keep their old code.
Thanks for your feedback.

**Comment by rds1983** [comment]
btw, small offtop question: why doesnt have MacOS platform DESKTOPGL macro?

**Comment by cra0zy** [comment]
> btw, small offtop question: why doesnt have MacOS platform DESKTOPGL macro?

MACOS platform is using Xamarin.Mac and native mac windowing and input, DESKTOPGL platform is using SDL. You can kind of compare that to WindowsDX and DesktopGL.

**Comment by rds1983** [comment]
I see, thanks.

**Comment by rds1983** [comment]
While commit is still there, I've commited some code. It's mostly refactoring and making generated code more cleaner and easier to read. But it is also some bug fixes.

**Comment by tomspilman** [comment]
@KonajuGames @dellis1972 - We need to make a final pass on this and decide on merging it.

It would only affect GL for now, but ideally after some performance testing and running it thru our unit tests we can then use it across all platforms.

**Comment by mrhelmut** [comment]
I gave this PR a test.

The good: it works and is faster than the original ```System.Drawing``` (and TONS faster than the Mono equivalent).

The bad: the alpha channel is wrong, I have all my alpha values having a ghostly color while it should be 100% transparent. I can't post any example right now and haven't looked into it yet (probably a channel swap).

**Comment by tomspilman** [comment]
Thanks @mrhelmut .

If you get a chance to provide @rds1983 an image that loads incorrect i'm sure we can figure out this issue.

**Comment by mrhelmut** [comment]
This should give wrong results when ```Texture2D.FromSteam()``` and rendering with ```SpriteBatch```: https://www.dropbox.com/s/8wgj3zs7fmc1un3/crop.png?dl=0

Maybe something to with premultiplied alpha?

**Comment by rds1983** [comment]
Hi @mrhelmut.
It seems that image doesnt have the alpha channel at all:
crop_desc

**Comment by cra0zy** [comment]
Looking at the image from Linux, it definitely has alpha.

**Comment by tomspilman** [comment]
@rds1983 - Ywah @cra0zy is right... it is a 32bit image.  Did you maybe right click "Save Image" on the preview image instead of downloading the file from DropBox?

**Comment by rds1983** [comment]
My bad, sorry. Now I downloaded it instead of just saving and alpha was there.

Anyway, the issue is related to the alpha-premultiplying when the texture is loaded from Texture2D.FromStream.
Different MG backends have different approaches there. I.e. WindowsDX backend does alpha-premultiplying. While DesktopGL doesnt. As this PR as aimed to replace DesktopGL, the alpha-premultiplying is not applied.
However you can still render the image correctly, using BlendState.NonPremultiplied.

**Comment by mrhelmut** [comment]
I believe that DesktopGL premultiplies alpha as well, because it renders correctly with ```BlendState.AlphaBlend``` on DesktopGL.

**Comment by tomspilman** [comment]
> Different MG backends have different approaches there.

Well the one I know is 100% correct is the Windows DirectX backend as I fixed it a few weeks ago.

 https://github.com/MonoGame/MonoGame/pull/5712

I recommend looking at that one and the docs on `Texture2D.FromStream` which I fixed.

It doesn't matter what any other backend does... Windows DirectX is correct and all others should be fixed to match it.

>  WindowsDX backend does alpha-premultiplying. 

It used to... which was wrong. It no longer does.

Correct XNA behavior is to not premultiply in `Texture2D.FromStream`.

However XNA does black out the color of any pixel with a zero in alpha.  You can see my implementation of that under DirectX here:

https://github.com/MonoGame/MonoGame/blob/develop/MonoGame.Framework/Graphics/Texture2D.DirectX.cs#L364

This behavior is why I think some people think it premultiplies...  it doesn't... it just does this.

**Comment by rds1983** [comment]
Thanks for the detailed clarification.
I've copied code that blacks out zero-alpha pixels to the new image loading code(Texture2D.OpenGL.cs) and now it renders @mrhelmut's image correctly using default SpriteBatch(which uses BlendState.AlphaBlend).

**Comment by rds1983** [comment]
A error(mac_build) I haven't seen before came out:
Utilities/ReflectionHelpers.Default.cs(19,35): error CS0308: The non-generic method 'Marshal.SizeOf(object)' cannot be used with type arguments

**Comment by tomspilman** [comment]
@rds1983 - Not your fault.  Rebase or merge the latest changes from the develop branch.

**Comment by tomspilman** [comment]
@rds1983 - Something weird is going on it shows a conflict now and 160 file changes.  Maybe it is how you rebased/merged?

**Comment by rds1983** [comment]
I took latest develop, switched to sysdraw(this PR branch) and did 'git rebase develop', made sure it compiles and runs, then did 'git push origin sysdraw --force'. Something went wrong.

**Comment by rds1983** [comment]
I forgot to sync mine develop with upstream's develop. It had been fixed now.

**Comment by mrhelmut** [comment]
Looks good to me. 👍 
Begone System.Drawing!

**Comment by tomspilman** [comment]
Great.... thanks @rds1983 and @mrhelmut !

Next step...

We should get the unit tests running under GL (https://github.com/MonoGame/MonoGame/pull/5587) so we can test these changes further.

Also replacing the other implementations with this one is fairly high priority.  My only concern is how fast this new implementation is vs native code.  I know on Switch i'm using a copy of STB for one project to avoid the current super slow C# png loader.  I hope to test this new loader on Switch to prove its performance.

**Pull request merged**